sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -611,6 +611,29 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Can't prune the columns on LeafNode
     case p @ Project(_, _: LeafNode) => p
 
+    // If the current project or the child references to nested fields, we can substitute them
```

Comment:
Can we please move this into a separate rule?

Reply:
I feel it should be part of `ColumnPruning`. 

Do you mean move this particular rule out or move the entire `ColumnPruning` into a new file? If the later, I chatted with @gatorsmile, and he felt moving the entire `ColumnPruning` out will lost the history which is not ideal.